bindings/java/src/integration/com/apple/foundationdb/CycleMultiClientIntegrationTest.java

```diff
@@ -0,0 +1,140 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.apple.foundationdb;
+
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import com.apple.foundationdb.tuple.Tuple;
+
+import org.junit.jupiter.api.Assertions;
+
+/**
+ * Setup: Generating a cycle 0 -> 1 -> 2 -> 3 -> 0, its length is 4
+ * Process: randomly choose an element, reverse 2nd and 4rd element, considering the chosen one as the 1st element.
+ * Check: verify no element is lost or added, and they are still a cycle.
+ * 
+ * This test is to verify the atomicity of transactions. 
+ */
+public class CycleMultiClientIntegrationTest {
+	public static final MultiClientHelper clientHelper = new MultiClientHelper();
+
+	private static final int len = 4;
+	private static List<String> expected = new ArrayList<>(Arrays.asList("0", "1", "2", "3"));
+
+	public static void main(String[] args) throws Exception {
+		FDB fdb = FDB.selectAPIVersion(630);
+		setupThreads(fdb);			
+		Collection<Database> dbs = clientHelper.openDatabases(fdb); // the clientHelper will close the databases for us
+		System.out.print("Starting tests\n");
+		setup(dbs);
+		System.out.print("Start processing\n");
+		process(dbs);
+		System.out.print("Start validataing\n");
+		check(dbs);
+		System.out.print("Test finished\n");
+	}
+
+	private static synchronized void setupThreads(FDB fdb) {
+		int clientThreadsPerVersion = clientHelper.readClusterFromEnv().length;
+		fdb.options().setClientThreadsPerVersion(clientThreadsPerVersion);
+		System.out.printf("thread per version is %d\n", clientThreadsPerVersion);
+		fdb.options().setExternalClientDirectory("/var/dynamic-conf/lib");
+		fdb.options().setTraceEnable("/tmp");
+		fdb.options().setKnob("min_trace_severity=5");
+	}
+
+	private static void setup(Collection<Database> dbs) {
+		// 0 -> 1 -> 2 -> 3 -> 0
+		for (int k = 0; k < len; k++) {
+			String key = Integer.toString(k);
+			String value = Integer.toString((k + 1) % len);
+
+			for (Database db : dbs) {
+				db.run(tr -> {
+					tr.set(Tuple.from(key).pack(), Tuple.from(value).pack());
+					return null;
+				});
+			}
+		}
+	}
+
+	private static void process(Collection<Database> dbs) {
+		for (int i = 0; i < len; i++) {
+			int k = ThreadLocalRandom.current().nextInt(len);
+			String key = Integer.toString(k);
+
+			for (Database db : dbs) {
```

Comment:
Instead of running transactions against one database at a time, spawn N threads per database, and run the loops inside those threads.  (As I understand it, the FDB client API is threadsafe from the application's perspective, so multiple threads can use the same instance of `db` without further synchronization.)